src/Libraries/CoreNodes/List.cs

```diff
@@ -235,6 +236,87 @@ public static IDictionary Deconstruct(IList list)
         }
 
         /// <summary>
+        ///     Sort list based on its keys
+        /// </summary>
+        /// <param name="list">list to be sorted</param>
+        /// <param name="keys">list of keys</param>
+        /// <returns name="sorted list">sorted list</returns>
+        /// <returns name="sorted keys">sorted keys</returns>
+        /// <search>sort;key</search>
+        [MultiReturn(new[] { "sorted list", "sorted keys" })]
+        public static IDictionary SortByKey(IList list, IList keys)
+        {
+            if (list == null || keys == null)
+                return null;
+
+            var containsSublists = keys.Cast<object>().Any(key => key is IList || key is ICollection);
+            if (containsSublists)
+            {
+                throw new ArgumentException(Resources.InvalidKeysErrorMessage);
+            }
+
+            if (list.Count != keys.Count)
+                return null;
+
+            var pairs = list.Cast<object>()
+                    .Zip(keys.Cast<object>(), (item, key) => new { item, key });
+
+            var numberKeyPairs = pairs.Where(pair => pair.key is double || pair.key is int || pair.key is float);
+            // We don't use Except, because Except doesn't return duplicates.
+            var keyPairs = pairs.Where(
+                pair =>
+                    !numberKeyPairs.Any(
+                        numberPair => numberPair.item == pair.item && numberPair.key == pair.key));
+
+            // Sort.
+            numberKeyPairs = numberKeyPairs.OrderBy(pair => Convert.ToDouble(pair.key));
+            keyPairs = keyPairs.OrderBy(pair => pair.key);
+
+            // First items with number keys, then items with letter keys.
+            var sortedPairs = numberKeyPairs.Concat(keyPairs);
+
+            var sortedList = sortedPairs.Select(x => x.item);
```

Comment:
I think it might be best to convert ToList() at this point before returning, this is done with the group by nodes and other list nodes return lists not ienums.